File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int main(void)
{
    long credit = get_long("Number: ");

    int sum = 0;
    bool finish_looping = false;
    int counter = 1;
    while (!finish_looping)
    {
        long digit = credit % (long) pow(10, counter);
        if (digit == credit)
        {
            finish_looping = true;
        }

        digit = digit / (long) pow(10, counter - 1);

        if (counter % 2 == 0)
        {
            digit = digit * 2;

            if (digit >= 10)
            {
                sum += 1;
                sum += digit % 10;
            }
            else
            {
                sum += digit;
            }
        }
        else
        {
            sum += digit;
        }

        counter++;
    }

    int credit_length = counter - 1;
    bool has_valid_sum = sum % 10 == 0;
    int first_two_digits = credit / (long) pow(10, credit_length - 2);
    int first_digit = first_two_digits / 10;

    if (has_valid_sum && credit_length == 15 && (first_two_digits == 34 || first_two_digits == 37))
    {
        printf("AMEX\n");
        return 0;
    }
    else if(has_valid_sum && credit_length == 16 && (first_two_digits == 51 || first_two_digits == 52 ||
                                                      first_two_digits == 53 || first_two_digits == 54 ||
                                                      first_two_digits == 54))
    {
        printf("MASTERCARD\n");
        return 0;
    }

    else if(has_valid_sum && (credit_length == 13 || credit_length == 16) && first_digit == 4)
    {
        printf("VISA\n");
        return 0;
    }

    printf("INVALID\n");
}

File: pset6/credit/credit.py
from cs50 import get_string

number = get_string("Number: ")

digits_sum = 0

for i in range(len(number) - 2 , -1, -2):
    digit = int(number[i]) * 2
    if digit >= 10:
        digits_sum += 1
        digits_sum += digit % 10
    else:
        digits_sum += digit

for i in range(len(number) - 1, -1, -2):
    digits_sum += int(number[i])

valid_sum = digits_sum % 10 == 0;
if valid_sum and len(number) == 15 and int(number[0:2]) in [34, 37]:
    print("AMEX")
elif valid_sum and len(number) == 16 and int(number[0:2]) in [51, 52, 53, 54, 55]:
    print("MASTERCARD")
elif valid_sum and len(number) in [13, 16] and int(number[0:1]) == 4:
    print("VISA")
else:
    print("INVALID")